### ISSUE TITLE
Proposal for removal of relationship entities from base specification

### ISSUE THREAD

**Issue by gr0uch** [json-api/json-api#623]
Relationship entities should not be a mandatory part of the specification, instead they should be deprecated in favor of the official patch extension. From my understanding, the reason why they exist in the base spec is to provide push/pull operations on `linkage` arrays. The patch extension provides this, and more (add/remove ops on any array, not just to-many links), and there is no reason why the patch extension needs a concept of relationship entities. A post/delete request to push/pull on a to-many `linkage` object is functionally equivalent between the base spec and patch extension. In the case of to-one `linkage` objects, relationship entity updates are functionally equivalent with a patch request on the resource itself.

**Current situation**:
- 3 (!) different ways to update a to-one `linkage` object: patch on resource, patch on relationship entity, patch extension with `replace` op.
- 7 (!!!) different ways to update a to-many `linkage` object: patch on resource, patch on relationship entity, post on relationship entity, delete on relationship entity, patch extension with `replace` op, patch extension with `add` op, patch extension with `remove` op.
- 3 update payload formats: patch on resource, [verb] on relationship entity, patch extension.

**Advantages** of removing relationship entities from the base spec:
- Less to implement for servers and clients, and one less concept to deal with.
- Instead of multiple ways to do the same thing, just do it consistently.
- No more awkward `/links/` routes.

**Disadvantages** of removing relationship entities:
- No more push/pull on linkage arrays in the base spec (only replacement, and if they need more they can use the patch extension).

**Result of proposal**:
- 2 ways to update a to-one `linkage` object: patch on resource, patch extension with `replace` op.
- 4 ways to update a to-one `linkage` object: patch on resource, patch extension with `replace` op, patch extension with `add` op, patch extension with `remove` op.
- 2 update payload formats: patch on resource, patch extension.

Note that this simplifies relationship updates in the base spec into just patching the resource. Removing relationship entities would greatly simplify things.

**Comment by patrykorwat** [comment]
I might be not so radical, some people may have already make use of this feature. I'm for creating an extension to which those methods will be moved.

**Comment by gr0uch** [comment]
I know that the timing for this proposal is bad on the eve of a 1.0 release of JSON API, but I strongly feel that it does not belong as part of the base spec since it adds complexity in many places for just one additional feature. However, other breaking changes have been merged recently such as the `attributes` object, so I think it's still fair game to remove relationship entities at this point.

**Comment by ethanresnick** [comment]
In my experience, push/pull on linkage arrays is valuable not only for preventing race conditions (which may be somewhat of a theoretical concern for many APIs), but also because it's _incredibly_ convenient, in that it allows clients to add something to a relationship without having to request the containing resource first. Therefore, I think implementors of the base spec should have an easy way to add support for push/pull on linkage arrays. That doesn't preclude the feature existing as an extension, per se, but I think it does preclude it existing only as part of the patch extension—which is a very different animal than the base spec, and not easy to "tack on" support for.

**Comment by gr0uch** [comment]
Post/delete on a relationship entity only provides a partial solution to the more generic problem of operations on an array. If its a truly necessary feature for an API to have, it should implement the patch extension for a generic solution, rather than relationship entities which exist solely as a special case for updating to-many relationships.

I don't think its a big deal breaker to not be able to modify a link without requesting the resource in the base spec, it just makes it slightly more inconvenient. I think dealing relationship entities is more inconvenient, and the issue could be solved with the patch extension. 

If the spec maintainers agree, I can put together a PR for this.

**Comment by tkellen** [comment]
I apologize that we did not respond to this sooner. You're right that relationships can be managed in a more general way with JSON-PATCH. That said, it greatly lowers the barrier of entry for implementations and consumers to support them. Another way of putting it would be to say that it's possible to have a nice API without using JSON-PATCH, but very difficult to have one without support for relationships.

Now that the specification is now at 1.0, we cannot remove them. I appreciate the feedback and I understand the heart of where you're coming from, but this is a tradeoff we've made with open eyes.
